Providers/Profile.swift

```diff
@@ -898,55 +946,66 @@ public class BrowserProfile: Profile {
-                go.upon({ res in self.endSyncing() })
-
-                return go
+        private func sync(label: EngineIdentifier, function: SyncFunction) -> SyncResult {
+            return syncSeveral((label, function)) >>== { statuses in
+                let engineStatus = statuses.filter({ $0.0 == label }).first!
+                return deferMaybe(engineStatus.1)
             }
-
-            log.warning("No account; can't sync.")
-            return nil
         }
 
         /**
-         * Runs the single provided synchronization function and returns its status.
+         * Runs each of the provided synchronization functions with the same inputs.
+         * Returns an array of IDs and SyncStatuses the same length as the input.
          */
-        private func sync(label: EngineIdentifier, function: (SyncDelegate, Prefs, Ready) -> SyncResult) -> SyncResult {
-            return self.withSyncInputs(label, function: function) ??
-                   deferMaybe(.NotStarted(.NoAccount))
+        private func syncSeveral(synchronizers: (EngineIdentifier, SyncFunction)...) -> Deferred<Maybe<[EngineStatus]>> {
+            return syncSeveral(synchronizers)
         }
 
         /**
          * Runs each of the provided synchronization functions with the same inputs.
          * Returns an array of IDs and SyncStatuses the same length as the input.
          */
-        private func syncSeveral(synchronizers: (EngineIdentifier, SyncFunction)...) -> Deferred<Maybe<[(EngineIdentifier, SyncStatus)]>> {
-            typealias Pair = (EngineIdentifier, SyncStatus)
-            let combined: (SyncDelegate, Prefs, Ready) -> Deferred<Maybe<[Pair]>> = { delegate, syncPrefs, ready in
+        private func syncSeveral(synchronizers: [(EngineIdentifier, SyncFunction)]) -> Deferred<Maybe<[EngineStatus]>> {
+
+            syncLock.lock()
+            defer { syncLock.unlock() }
+            let go: Deferred<Maybe<[EngineStatus]>>
+            let requestedLabels = synchronizers.map { $0.0 }
+            if !beginSyncing() {
+                log.info("Already syncing something. Will queue for later: \(requestedLabels)")
+                go = currentSync! >>== { justSynced in self.syncRemaining(synchronizers, except: justSynced) }
+            } else {
+                log.info("Beginning a sync of \(requestedLabels)")
+                go = deferDispatchAsync(syncQueue, f: { self.syncWith(synchronizers) })
+            }
+
+            // By the time currentSync finishes, it may not be the last sync, so 
+            // endSyncingMaybe will check, and end if necessary.
+            go.upon({ res in
+                log.info("Ending current sync. \(requestedLabels)")
+                // In the case of failure, then sync status is NotStarted.
+                // This is because we don't know which ones sync function failed.
+                self.endSyncingMaybe(res.successValue ?? synchronizers.map { ($0.0, .NotStarted(.Offline)) })
```

Comment:
I don't think this is quite correct. This means that we'll never actually fail to sync, in the sense of filling the `Deferred` with an error; instead we'll always report that we haven't started the next set. I think you need a failure version of `endSyncingMaybe`/`endSyncing` so you can reject the chain.

But if you think about it, that's exactly what a `Deferred<Result>` is — it's a pending success or failure. The only reason this entire chunk of code isn't just chaining `go` into `syncTerminal` is that we might be inserting another link in the chain, and `Deferred` isn't set up to unlink the previous!

My proposal, then: `endSyncingMaybe` should take the `Deferred` that just finished. If it's the last one, fill `syncTerminal` with its values.

(Modulo one thing, which I don't think you've handled here; we'll see when I've finished the review!)